hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto

```diff
@@ -528,6 +529,12 @@ enum StorageTypeProto {
     RAM_DISK = 4;
 }
 
+enum BucketTypeProto {
```

Comment:
Thanks for the suggestion and for the patch @aryangupta1998 . I already shared my concern offline, but let me repeat m y arguments here in public:

While I really like the concept of having different type of buckets (or more preciously: different type of behavior for different buckets), I am not sure if the best representation is introducing a new bucket "type".

We already have a custom `Map<String,String>` field, which (today) is used to store the flags about FSO or non-FSO behavior.

Having both the custom fields (aka. bucket properties/behaviors) and the bucket type seems to be confusing for me from usability point of view.

I have no problem with using bucket type on CLI but wouldn't be enough to implement it only on the client side and use the good old `Map<String,String>` properties? 


It would be more backward and forward compatible (using enum makes it harder to introduce new values without compatibility issues.)

What do you think?

Reply:
@elek , Can you please elaborate on this point. Enums are not inherently backward incompatible.  Using strings has no other advantages either.

`using enum makes it harder to introduce new values without compatibility issues.`

Reply:
Also, the plan is not to have both the `Map<String,String>` as well as the enum together. The `Map<String,String>` will not be used once the enum has been introduced.

Reply:
> Also, the plan is not to have both the `Map<String,String>` as well as the enum together. The Map<String,String> will not be used once the enum has been introduced.

I am against removing the `Map<String,String>` as it can be used not only for storing bucket types but storing any other metadata. Also, it's an incompatible change if we remove it.

And if we need to keep it: wouldn't it make more sense to use it to store the bucket type (as we do it today?).

Again: I fully support to use bucket type as a concept: talk about multiple bucket types, support bucket type argument in cli. I just try to understand the arguments to introduce it a new protobuf level concept instead of using existing metadata capabilities.

Reply:
Thanks a lot @marton for the comments.

> I am against removing the Map<String,String> as it can be used not only for storing bucket types but storing any other metadata. Also, it's an incompatible change if we remove it.

To make it clear about the`Map<String,String>` entity - this patch will not remove `KeyValue` attribute from the BucketInfo and it will remain the same. Any upcoming feature or behaviour can still make use of this `KeyValue` attribute and build it.

_OmClientProtocol.proto_
```
    message BucketInfo {
    .....
    repeated hadoop.hdds.KeyValue metadata = 7;
    .....
+ optional BucketLayoutProto bucketLayout = 18;
```

This patch is moving the bucket layout information stored in the **Map to an enum attribute.**

> Also, it's an incompatible change if we remove it.

IIUC, since the FSO(prefix) feature is not released in any of the Ozone version we have still the opportunity to refactor/redesign the proto part. Please let me know if I missed anything.

Following are the points while evaluating `enum` over `Map<String, String>`.

- enum has options to limit the set of valid values and will validate the input efficiently. Please refer, https://picocli.info/#_enum_types
- protobuf defaults to the enum constant at ordinal 0 and not worrying about null checks in a programatic way(easily avoid NPE cases). Not only from the commandline, while reading back the old data from DB, where the bucket layout will not exists in DB.
- according to me enum makes code more maintainable and easy to read

I'm completely open for a better approach. With map, we have to do all the above in a programatic way and its doable. Please let me know your feedback. Thanks again!

Reply:
Thanks for the reply @elek, the proposal is not to remove the Map<String,String>. That will remain as it is.

Reply:
@elek,
If no more comments by EOD IST then I will commit this. If any more concerns, I can address them in later jira.

I would like to get this in, as this is blocking further jira's progress.
Thanks again for the reviews.

Reply:
Closing the discussion thread, thanks everyone!